packages/react-core/src/helpers/Popper/Popper.tsx

```diff
@@ -40,16 +85,84 @@ export const Popper: React.FunctionComponent<PopperProps> = ({
   popperMatchesTriggerWidth = true,
   direction = 'down',
   position = 'left',
+  placement,
   appendTo = () => document.body,
   zIndex = 9999,
-  isVisible = true
+  isVisible = true,
+  positionModifiers,
+  distance = 0,
+  onMouseEnter,
+  onMouseLeave,
+  onFocus,
+  onBlur,
+  onDocumentClick,
+  onTriggerClick,
+  onTriggerEnter,
+  onPopperClick,
+  onDocumentKeyDown,
+  enableFlip = true,
+  flipBehavior = 'flip'
+  // boundary = 'viewport',
 }) => {
   const [triggerElement, setTriggerElement] = React.useState(null);
   const [popperElement, setPopperElement] = React.useState(null);
   const [ready, setReady] = React.useState(false);
-  React.useEffect(() => setReady(true));
-  const getPlacement = () =>
-    `${direction === 'up' ? 'top' : 'bottom'}-${position === 'right' ? 'end' : 'start'}` as Placement;
+  const onDocumentClickCallback = React.useCallback(event => onDocumentClick(event, triggerElement), [
+    isVisible,
+    triggerElement,
+    onDocumentClick
+  ]);
+  React.useEffect(() => {
+    setReady(true);
+  }, []);
+  React.useEffect(() => {
+    onMouseEnter && triggerElement && triggerElement.addEventListener('mouseenter', onMouseEnter);
+    onMouseLeave && triggerElement && triggerElement.addEventListener('mouseleave', onMouseLeave);
+    onFocus && triggerElement && triggerElement.addEventListener('focus', onFocus);
+    onBlur && triggerElement && triggerElement.addEventListener('blur', onBlur);
+    onTriggerClick && triggerElement && triggerElement.addEventListener('click', onTriggerClick, false);
+    onTriggerEnter && triggerElement && triggerElement.addEventListener('keydown', onTriggerEnter, false);
+    onPopperClick && popperElement && popperElement.addEventListener('click', onPopperClick, false);
+    onDocumentClick && triggerElement && document.addEventListener('click', onDocumentClickCallback, false);
+    onDocumentKeyDown && triggerElement && document.addEventListener('keydown', onDocumentKeyDown, false);
+    return () => {
+      onMouseEnter && triggerElement && triggerElement.removeEventListener('mouseenter', onMouseEnter);
+      onMouseLeave && triggerElement && triggerElement.removeEventListener('mouseleave', onMouseLeave);
+      onFocus && triggerElement && triggerElement.removeEventListener('focus', onFocus);
+      onBlur && triggerElement && triggerElement.removeEventListener('blur', onBlur);
+      onTriggerClick && triggerElement && triggerElement.removeEventListener('click', onTriggerClick, false);
+      onTriggerEnter && triggerElement && triggerElement.removeEventListener('keydown', onTriggerEnter, false);
+      onPopperClick && popperElement && popperElement.removeEventListener('click', onPopperClick, false);
+      onDocumentClick && triggerElement && document.removeEventListener('click', onDocumentClickCallback, false);
+      onDocumentKeyDown && triggerElement && document.removeEventListener('keydown', onDocumentKeyDown, false);
```

Comment:
This code isn't very DRY. How about a helper function?
```js
function addEventListener(listener, element, event) {
  if (listener && element) {
    element.addEventListener(event, listener);
  }
}
```

You don't have to worry about the `useCapture = false` since [it's the default.](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener)